build-caching-staging-1/main.tf

```diff
@@ -0,0 +1,204 @@
+variable "cache_size_mb" {
+  default = 1848
+}
+
+variable "env" {
+  default = "staging"
+}
+
+variable "github_users" {}
+
+variable "index" {
+  default = 1
+}
+
+variable "machine_type" {
+  default = "custom-1-2048"
+}
+
+variable "project" {
+  default = "travis-staging-1"
+}
+
+variable "region" {
+  default = "us-central1"
+}
+
+variable "syslog_address_com" {}
+
+variable "travisci_net_external_zone_id" {
+  default = "Z2RI61YP4UWSIO"
+}
+
+terraform {
+  backend "s3" {
+    bucket         = "travis-terraform-state"
+    key            = "terraform-config/build-caching-staging-1.tfstate"
+    region         = "us-east-1"
+    encrypt        = "true"
+    dynamodb_table = "travis-terraform-state"
+  }
+}
+
+provider "google" {
+  project = "${var.project}"
+  region  = "${var.region}"
+}
+
+provider "google-beta" {
+  project = "${var.project}"
+  region  = "${var.region}"
+}
+
+provider "aws" {}
+
+data "template_file" "nginx_conf_d_default" {
+  template = "${file("${path.module}/nginx-conf.d-default.conf.tpl")}"
+
+  vars {
+    max_size = "${replace("${var.cache_size_mb * 0.9}", "/\\..*/", "")}m"
+  }
+}
+
+data "template_file" "cloud_config" {
+  template = "${file("${path.module}/cloud-config.yml.tpl")}"
+
+  vars {
+    assets               = "${path.module}/../assets"
+    here                 = "${path.module}"
+    nginx_conf_d_default = "${data.template_file.nginx_conf_d_default.rendered}"
+    syslog_address       = "${var.syslog_address_com}"
+
+    docker_env = <<EOF
+export TRAVIS_DOCKER_DISABLE_DIRECT_LVM=1
+EOF
+
+    github_users_env = <<EOF
+export GITHUB_USERS='${var.github_users}'
+EOF
+
+    squignix_env = <<EOF
+### squignix.env
+${file("${path.module}/squignix.env")}
+
+### in-line
+export SQUIGNIX_CACHE_SIZE=${format("%dk", var.cache_size_mb * 1024)}
+EOF
+  }
+}
+
+resource "google_compute_firewall" "build_cache" {
+  name      = "build-cache"
+  network   = "main"
+  direction = "INGRESS"
+
+  allow {
+    protocol = "tcp"
+    ports    = ["80", "8080", "11371"]
+  }
+
+  target_tags = ["build-cache"]
+}
+
+resource "google_compute_address" "build_cache" {
+  name    = "build-cache-1"
+  region  = "${var.region}"
+  project = "${var.project}"
+}
+
+resource "aws_route53_record" "build_cache" {
+  zone_id = "${var.travisci_net_external_zone_id}"
+  name    = "build-cache-${var.env}-${var.index}.gce-${var.region}.travisci.net"
+  type    = "A"
+  ttl     = 5
+
+  records = ["${google_compute_address.build_cache.address}"]
+}
+
+resource "google_compute_health_check" "build_cache" {
+  name = "build-cache-health-check"
+
+  timeout_sec        = 3
+  check_interval_sec = 5
+
+  http_health_check {
+    request_path = "/__squignix_health__"
+  }
+}
+
+resource "google_compute_backend_service" "build_cache" {
+  name        = "build-cache-backend"
+  description = "backend servers for build cache"
+  port_name   = "http"
+  protocol    = "HTTP"
+  timeout_sec = 10
+  enable_cdn  = false
```

Comment:
Maybe we could try the CDN magic in the future! (unless there's a reason not to)

Reply:
I'm happy to keep looking into this in the future, yep! I hesitated to glom it onto this work because of the problems associated with cache invalidation. The solution here is extremely simplified and doesn't deal with cache invalidation outside of capping the `tmpfs` size.